### PR TITLE
Remove panic() calls

### PR DESCRIPTION
**Pull request by Bslabe123**
BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/7174

### CODE REVIEW

File: codegen/kuberesource/crd.go [hunk 1]
```diff
@@ -175,7 +177,8 @@ func CustomResourceDefinition(
 		// hashstructure of the crd spec:
 		specHash, err := hashstructure.Hash(crd.Spec, nil)
 		if err != nil {
-			panic(err)
+			contextutils.LoggerFrom(nil).DPanic(err)
```
Comment (ben-taussig-solo) [inline codegen/kuberesource/crd.go:180]: Given that this is in codegen, I think we might prefer to panic here. I think codegen failing loudly is preferable to potentially returning an incorrect or undesirable result

File: contrib/README.md [hunk 2]
```diff
@@ -19,7 +19,8 @@ const (
 var Sets = func() model.CustomTemplates {
 	templateContents, err := templatesBox.FindString(SetCustomTemplatePath)
 	if err != nil {
-		panic(err)
```
Comment (ben-taussig-solo) [inline contrib/README.md:22]: I think we want to revert this -- I think the intended behavior is that we panic if we can't find the template in the specified location

File: contrib/codegen/funcs/funcs.go [hunk 3]
```diff
@@ -6,6 +6,8 @@ import (
 	"strings"
 	"text/template"
 
+	"github.com/solo-io/go-utils/contextutils"
```
Comment (ben-taussig-solo) [inline contrib/codegen/funcs/funcs.go:9]: I think we can revert the changes here -- I don't think it's an issue if codegen panics

File: contrib/codegen/templates/sets/sets.gotmpl [hunk 4]
```diff
@@ -136,7 +137,7 @@ func (s *{{ $kindLowerCamel }}Set) Insert(
         {{ $kindLowerCamel }}List ...*{{ $import_prefix }}.{{ $resource.Kind }},
 ) {
     if s == nil {
-        panic("cannot insert into nil set")
+        contextutils.LoggerFrom(nil).DPanic("cannot insert into nil set")
```
Comment (ben-taussig-solo) [inline contrib/codegen/templates/sets/sets.gotmpl:140]: Same as above -- I think we can revert the changes in this file because it isn't problematic for codegen to panic

**Review by ben-taussig-solo — COMMENTED** [review]
Left a couple of comments -- Essentially, I question whether we should be making changes to codegen, as I think it might be preferable for codegen to panic in the case of bad input.

File: contrib/custom_top_level_templates.go [hunk 5]
```diff
@@ -96,10 +98,12 @@ const (
 // Returns the template for generating input snapshots.
 func InputSnapshot(params SnapshotTemplateParameters) model.CustomTemplates {
 	templateContentsBytes, err := ioutil.ReadFile(templatesDir + InputSnapshotCustomTemplatePath)
+	var templateContents string
 	if err != nil {
-		panic(err)
+		contextutils.LoggerFrom(nil).DPanic(err)
+		templateContents = ""
```
Comment (sam-heilbron) [inline contrib/custom_top_level_templates.go:104]: This sets the templatecontents, and then the next line (outside the conditional) overwrite it. Do we need to do this or should we be returning an error here?

File: contrib/codegen/templates/sets/sets.gotmpl [hunk 6]
```diff
@@ -6,6 +6,7 @@ import (
     {{ $import_prefix }} "{{ group_import_path $ }}"
 
     "github.com/rotisserie/eris"
+    "github.com/solo-io/go-utils/contextutils"
```
Comment (sam-heilbron) [inline contrib/codegen/templates/sets/sets.gotmpl:9]: Why is this import needed?

File: changelog/v0.23.5/remove-panic.yaml [hunk 7]
```diff
@@ -0,0 +1,6 @@
+changelog:
+  - type: FIX
+    issueLink: https://github.com/solo-io/gloo/issues/7174
+    resolvesIssue: true
```
Comment (sam-heilbron) [inline changelog/v0.23.5/remove-panic.yaml:4]: false perhaps? Since we'll need to pull in these changes to gloo, I assume this won't be true until we have made the change in all repos

File: pkg/events/watcher.go [hunk 8]
```diff
@@ -106,7 +108,7 @@ func (w *eventWatcher) Reconcile(ctx context.Context, request reconcile.Request)
 			return reconcile.Result{}, err
 		}
 	default:
-		panic("invalid event")
+		return reconcile.Result{}, eris.Errorf("invalid event")
```
Comment (sam-heilbron) [inline pkg/events/watcher.go:111]: Do you think it makes sense to include any context around the event type so that if it occurs we can more easily debug it?

**Comment by sam-heilbron** [comment]
@Sodman can you or someone from the Mesh team check out this to make sure we're on the same page about the implications?

**Review by ben-taussig-solo — APPROVED** [review]
Looks good to me. For context, I've pushed to keep panics in codegen, under the assumptions that panicking on bad input is desired behavior in codegen/tests/other code which doesn't operate on the application path.

**Comment by Sodman** [comment]
If I understand correctly, these changes only affect codegen, and not any runtime, right?

My take here is that I'd like it if codegen explodes loudly and abruptly if there's something wrong. If we are going this route though, I think we should still return early with an error though, as otherwise we'll probably just log a single line in what's already a very output-verbose cli command - I'm worried about codegen accidentally being "broken" but the CLI command suddenly returning a happy exit 0 status and the developer not noticing  / committing the code to a PR.

If we're trying to remove `panic()`'s to keep a linter happy or some such though, I'm sure we can change our codegen on the mesh side make targets to enable devlogger, which should still result in panics on our side when we run this if we want them.

File: contrib/custom_top_level_templates.go [hunk 9]
```diff
@@ -113,10 +117,14 @@ const (
 // Returns the template for generating input snapshots.
 func InputSnapshotManualBuilder(params SnapshotTemplateParameters) model.CustomTemplates {
 	templateContentsBytes, err := ioutil.ReadFile(templatesDir + InputSnapshotManualBuilderCustomTemplatePath)
+	var templateContents string
 	if err != nil {
-		panic(err)
+		if err != nil {
```
Comment (sam-heilbron) [inline contrib/custom_top_level_templates.go:122]: I think this duplicate error check is unnecessary

File: contrib/custom_top_level_templates.go [hunk 10]
```diff
@@ -136,7 +144,7 @@ func InputReconciler(params SnapshotTemplateParameters) model.CustomTemplates {
 	}
 	templateContentsBytes, err := ioutil.ReadFile(templatesDir + templatePath)
 	if err != nil {
-		panic(err)
+		contextutils.LoggerFrom(nil).DPanic(err)
```
Comment (sam-heilbron) [inline contrib/custom_top_level_templates.go:147]: Should we return error as well?

File: contrib/custom_top_level_templates.go (hunk 9 above)
Comment (EItanya) [reply contrib/custom_top_level_templates.go:122]: Same comment as above

File: contrib/custom_top_level_templates.go (hunk 10 above)
Comment (EItanya) [reply contrib/custom_top_level_templates.go:147]: Same comment

File: contrib/codegen/templates/sets/sets.gotmpl (hunk 6 above)
Comment (Bslabe123) [reply contrib/codegen/templates/sets/sets.gotmpl:9]: This was leftover from a previous change, fixed

**Comment by Bslabe123** [comment]
See https://solo-io-corp.slack.com/archives/CHAC8587L/p1663947139651859?thread_ts=1663946678.931739&cid=CHAC8587L
